provider/azure/internal/azureauth/interactive.go

```diff
@@ -0,0 +1,347 @@
+			p = tracer(p)
+			p = requestInspector(p)
+			return p
+		}
+	}
+}
+
+func createOrUpdateServicePrincipal(
+	client ad.ServicePrincipalsClient,
+	subscriptionId string,
+	clock clock.Clock,
+	newUUID func() (utils.UUID, error),
+) (servicePrincipalObjectId, password string, _ error) {
+	passwordCredential, err := preparePasswordCredential(clock, newUUID)
+	if err != nil {
+		return "", "", errors.Annotate(err, "preparing password credential")
+	}
+
+	servicePrincipal, err := client.Create(
+		ad.ServicePrincipalCreateParameters{
+			ApplicationID:       jujuApplicationId,
+			AccountEnabled:      true,
+			PasswordCredentials: []ad.PasswordCredential{passwordCredential},
+		},
+		nil, // abort
+	)
+	if err != nil {
+		if !isMultipleObjectsWithSameKeyValueErr(err) {
+			return "", "", errors.Trace(err)
+		}
+	} else {
+		return servicePrincipal.ObjectID, passwordCredential.Value, nil
+	}
+
+	// The service principal already exists, so we need to query
+	// its object ID, and fetch the existing password credentials
+	// to update.
+	servicePrincipal, err = getServicePrincipal(client)
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	if err := addServicePrincipalPasswordCredential(
+		client, servicePrincipal.ObjectID,
+		passwordCredential,
+	); err != nil {
+		return "", "", errors.Annotate(err, "updating password credentials")
+	}
+	return servicePrincipal.ObjectID, passwordCredential.Value, nil
+}
+
+func isMultipleObjectsWithSameKeyValueErr(err error) bool {
+	if err, ok := errorutils.ServiceError(err); ok {
+		return err.Code == "Request_MultipleObjectsWithSameKeyValue"
+	}
+	return false
+}
+
+func preparePasswordCredential(
+	clock clock.Clock,
+	newUUID func() (utils.UUID, error),
+) (ad.PasswordCredential, error) {
+	password, err := newUUID()
+	if err != nil {
+		return ad.PasswordCredential{}, errors.Annotate(err, "generating password")
+	}
+	passwordKeyUUID, err := newUUID()
+	if err != nil {
+		return ad.PasswordCredential{}, errors.Annotate(err, "generating password key ID")
+	}
+	startDate := clock.Now().UTC()
+	endDate := startDate.Add(passwordExpiryDuration)
+	return ad.PasswordCredential{
+		// TODO(axw) include the credential name?
+		CustomKeyIdentifier: []byte("juju-" + startDate.Format("20060102")),
+		KeyId:               passwordKeyUUID.String(),
+		Value:               password.String(),
+		StartDate:           startDate,
+		EndDate:             endDate,
+	}, nil
+}
+
+func addServicePrincipalPasswordCredential(
+	client ad.ServicePrincipalsClient,
+	servicePrincipalObjectId string,
+	passwordCredential ad.PasswordCredential,
+) error {
+	existing, err := client.ListPasswordCredentials(servicePrincipalObjectId)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	passwordCredentials := append(existing.Value, passwordCredential)
+	if _, err := client.UpdatePasswordCredentials(
+		servicePrincipalObjectId,
+		ad.PasswordCredentialsUpdateParameters{passwordCredentials},
+	); err != nil {
```

Comment:
You could do without the `if` and just `return errors.Trace(err)` the error returned by `UpdatePasswordCredentials`.